File: impl/sdk/java/src/examples/java/io/resonatehq/examples/errorhandling/ErrorHandling.java
package io.resonatehq.examples.errorhandling;

import io.resonatehq.resonate.Context;
import io.resonatehq.resonate.Errors.ApplicationError;
import io.resonatehq.resonate.Resonate;
import io.resonatehq.resonate.Retry.Never;
import java.util.List;

/**
 * Error handling across the Resonate durability boundary.
 *
 * <p>When a durable function like {@code bar} fails, its exception does not propagate in-process the
 * way a normal call would. Resonate <b>encodes the failure, writes it to a durable promise, and
 * later decodes it</b> for whoever awaits the result -- the same process, a different worker (via
 * {@code ctx.rpc}), or a process that recovered after a crash.
 *
 * <p>The Java wire form carries a transport-safe {@code message} string plus a best-effort native
 * serialization of the original exception (the analogue of Python's pickle). So a same-runtime
 * awaiter that can load the class gets the <b>original type back</b> -- {@code foo} below catches
 * {@link UsernameTakenError} and {@link IllegalArgumentException} directly. When the serialized form
 * cannot round-trip, the awaiter falls back to a plain {@link ApplicationError} carrying the
 * message.
 *
 * <pre>{@code ./gradlew runExample -PmainClass=io.resonatehq.examples.errorhandling.ErrorHandling \
 *     -PexampleArgs="--mode run --error taken"}</pre>
 */
public final class ErrorHandling {
    private ErrorHandling() {}

    /** Raised when a username is already registered. A {@link RuntimeException} so it round-trips by type. */
    public static final class UsernameTakenError extends RuntimeException {
        public UsernameTakenError(String message) {
            super(message);
        }
    }

    public static String bar(Context ctx, String username, int age) {
        List<String> existingUsers = List.of("admin", "coder123", "python_fan");

        if (age < 0) {
            throw new IllegalArgumentException("Age cannot be negative.");
        }
        if (existingUsers.contains(username.toLowerCase())) {
            throw new UsernameTakenError("The username '" + username + "' is already in use.");
        }
        return "Registration successful for " + username + "!";
    }

    public static String foo(Context ctx, String username, int age, String mode) {
        try {
            String v = (String)
                    (mode.equals("run")
                            ? ctx.run(ErrorHandling::bar, username, age).await()
                            : ctx.rpc("bar", username, age).await());
            System.out.println("🎉 Success: " + v);
            return v;
        } catch (UsernameTakenError e) {
            // The original domain type is reconstructed across the boundary, so we can catch it
            // directly instead of parsing a message prefix.
            System.out.println("⚠️ Business Rule Error: " + e.getMessage());
            System.out.println("💡 Suggestion: Please try adding numbers to your username.");
            return null;
        } catch (IllegalArgumentException e) {
            System.out.println("💥 Validation Error: " + e.getMessage());
            return null;
        } catch (ApplicationError e) {
            throw new AssertionError(
                    "ApplicationError is the fallback used when the raised error cannot be deserialized."
                            + " Not applicable in this example.",
                    e);
        }
    }

    public static void main(String[] args) {
        String mode = argValue(args, "--mode", "run");
        String error = argValue(args, "--error", "none");

        String username;
        int age;
        if (error.equals("taken")) {
            username = "admin"; // Triggers UsernameTakenError
            age = 25;
        } else if (error.equals("value")) {
            username = "alice"; // Triggers IllegalArgumentException
            age = -5;
        } else {
            username = "alice"; // Successful path
            age = 25;
        }

        String url = System.getenv().getOrDefault("RESONATE_URL", "http://localhost:8001");
        Resonate r = Resonate.builder().url(url).retryPolicy(new Never()).build();
        // Registering both functions so RPC coordination works flawlessly.
        r.register(ErrorHandling::foo);
        r.register(ErrorHandling::bar);

        try {
            String id = "error-handling-" + System.nanoTime();
            r.run(id, ErrorHandling::foo, username, age, mode).result();
            r.get(id).join().done();
        } finally {
            r.stop().join();
        }
    }

    private static String argValue(String[] args, String flag, String fallback) {
        for (int i = 0; i + 1 < args.length; i++) {
            if (args[i].equals(flag)) {
                return args[i + 1];
            }
        }
        return fallback;
    }
}

File: impl/sdk/ts/examples/generator/error-handling.ts
// error-handling shows how a failure raised in a durable function reaches
// whoever awaits it.
//
// When a durable function like `bar` fails, its exception does not propagate
// in-process the way a normal throw would. Resonate **encodes the failure,
// writes it to a durable promise, and later decodes it** for whoever awaits the
// result -- possibly a different worker, or a process that recovered after a
// crash and never saw the original throw.
//
// The codec encodes an Error's `name`, `message`, and `stack`. On the far side
// the awaiter gets an `Error` whose `.name` is preserved, so you discriminate
// domain failures by name (TS has no pickle, so the concrete class/prototype is
// not reconstructed -- `instanceof` on a custom subclass will not match).
//
//   npx tsx examples/generator/error-handling.ts --mode run --error taken
//   npx tsx examples/generator/error-handling.ts --mode rpc --error value

import { type Context, Never, Resonate } from "../../src/index.js";

class UsernameTakenError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UsernameTakenError";
  }
}

async function bar(_ctx: Context, username: string, age: number): Promise<string> {
  const existingUsers = ["admin", "coder123", "python_fan"];
  if (age < 0) throw new RangeError("Age cannot be negative.");
  if (existingUsers.includes(username.toLowerCase())) {
    throw new UsernameTakenError(`The username '${username}' is already in use.`);
  }
  return `Registration successful for ${username}!`;
}

function* foo(ctx: Context, username: string, age: number, mode: string): Generator<any, void, any> {
  // Never on the leaf call: surface the failure on the first attempt. (The
  // generator engine defaults leaf functions to Exponential retry.)
  const opts = ctx.options({ retryPolicy: new Never() });
  try {
    const v: string = yield* mode === "run"
      ? ctx.run(bar, username, age, opts)
      : ctx.rpc<string>("bar", username, age, opts);
    console.log(`🎉 Success: ${v}`);
  } catch (err) {
    // The original type is NOT reconstructed across the boundary, but `name`
    // survives -- so discriminate on it rather than on `instanceof`.
    const e = err as Error;
    if (e.name === "UsernameTakenError") {
      console.log(`⚠️  Business Rule Error: ${e.message}`);
      console.log("💡 Suggestion: Please try adding numbers to your username.");
    } else if (e.name === "RangeError") {
      console.log(`💥 Validation Error: ${e.message}`);
    } else {
      throw err;
    }
  }
}

function flag(name: string, fallback: string): string {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 && process.argv[i + 1] ? process.argv[i + 1] : fallback;
}

const mode = flag("mode", "run");
const error = flag("error", "none");
const [username, age] = error === "taken" ? ["admin", 25] : error === "value" ? ["alice", -5] : ["alice", 25];

const resonate = new Resonate({ url: process.env.RESONATE_URL ?? "http://localhost:8001" });
resonate.register("foo", foo);
resonate.register("bar", bar); // registered so rpc-by-name can find it

try {
  const id = `error-handling-${Date.now()}`;
  await resonate.run(id, foo, username as string, age as number, mode);
} finally {
  await resonate.stop();
}
